gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+
+
+@_expand_all_product.register(Product)
+def _expand_all_product_common(node, self, index):
+    a = self(node.children[0], index)
+    b = self(node.children[1], index)
+    if isinstance(b, Sum) and any([i in b.free_indices for i in index]):
+        return Sum(self(Product(a, b.children[0]), index),
+                   self(Product(a, b.children[1]), index))
+    elif isinstance(a, Sum) and any([i in a.free_indices for i in index]):
+        return Sum(self(Product(a.children[0], b), index),
+                   self(Product(a.children[1], b), index))
+    else:
+        return node
+
+
+def expand_all_product(node, index):
+    mapper = MemoizerArg(_expand_all_product)
+    return mapper(node, index)
+
+
+def collect_terms(node, node_type):
+    """Recursively collect all children into a list from :param:`node`
+    and its children of class :param:`node_type`.
+
+    :param node: root of expression
+    :param node_type: class of node (e.g. Sum or Product)
+    :return: list of all terms
+    """
+    from collections import deque
+    terms = []  # collected terms
+    queue = deque([node])  # queue of children nodes to process
+    while queue:
+        child = queue.popleft()
+        if isinstance(child, node_type):
+            queue.extendleft(reversed(child.children))
+        else:
+            terms.append(child)
+    return tuple(terms)
+
+
+def flatten_sum(node, argument_indices):
+    """
+    factorise :param:`node` into sum of products, group factors of each product
+    based on its dependency on index:
+    1) nothing (key = 0)
+    2) i (key = 1)
+    3) (i)j (key = j)
+    4) (i)k (key = k)
+    ...
+    5) (i)jk (key = 2)
+    :param node: root of expression
+    :param index: tuple of argument (linear) indices
+    :return: dictionary to list of factors
+    """
+    monos = collect_terms(node, Sum)
+    result = []
+    arg_ind_set = set(argument_indices)
+    for mono in monos:
+        d = OrderedDict()
+        for i in [0, 1, 2] + list(argument_indices):
+            d[i] = list()
+        for factor in collect_terms(mono, Product):
+            fi = factor.free_indices
+            if not fi:
+                d[0].append(factor)
+            else:
+                ind_set = set(fi) & arg_ind_set
+                if len(ind_set) > 1:
+                    # Aijk
+                    d[2].append(factor)
+                elif len(ind_set) == 0:
+                    # Ai
+                    d[1].append(factor)
+                else:
+                    # Aij
+                    d[ind_set.pop()].append(factor)
+            # elif j in fi and k in fi:
+            #     d[2].append(factor)
+            # elif j in fi:
+            #     d[j].append(factor)
+            # elif k in fi:
+            #     d[k].append(factor)
+            # else:
+            #     d[1].append(factor)
+        for i in d:
+            d[i] = tuple(d[i])
+        result.append(d)
+    return tuple(result)
```

Comment:
Will have a think about it. I got another data structure which is very similar, but as a list of lists (instead of list of dicts), I really should just use one, and also keep it through the run, instead of collapsing it and expanding again during recursion.